staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/fairqueuing.go

```diff
@@ -0,0 +1,633 @@
+	return pkt, ok
+}
+
+func (qs *QueueSetImpl) roundrobinqueue() int {
+	// TODO(aaron-prindle) verify this is modified on quiesce...
+	qs.robinIdx = (qs.robinIdx + 1) % len(qs.queues)
+	return qs.robinIdx
+}
+
+// selectQueue selects the minimum virtualfinish time from the set of queues
+// the starting queue is selected via roundrobin
+// TODO(aaron-prindle) verify that the roundrobin usage is correct
+// unsure if the code currently prioritizes the correct queues for ties
+func (qs *QueueSetImpl) selectQueue() *Queue {
+	minvirfinish := math.Inf(1)
+	var minqueue *Queue
+	var minidx int
+	for range qs.queues {
+		// TODO(aaron-prindle) how should this work with queue deletion?
+		idx := qs.roundrobinqueue()
+		queue := qs.queues[idx]
+		if len(queue.Packets) != 0 {
+			curvirfinish := queue.GetVirtualFinish(0, qs.estimatedServiceTime)
+			if curvirfinish < minvirfinish {
+				minvirfinish = curvirfinish
+				minqueue = queue
+				minidx = idx
+			}
+		}
+	}
+	qs.robinIdx = minidx
+	return minqueue
+}
+
+// AddQueues adds additional queues to the QueueSetImpl
+// the complementary DeleteQueues is not an explicit fxn as queue deletion requires draining
+// the queues first, queue deletion is done for the proper cases
+// in the the FinishPacket function
+func (qs *QueueSetImpl) addQueues(n int) {
+	for i := 0; i < n; i++ {
+		qs.queues = append(qs.queues, &Queue{
+			Packets: []*Packet{},
+		})
+	}
+}
+
+// ===========================================================================
+// ===========================================================================
+
+// Quiesce controls whether this system is quiescing.  Passing a
+// non-nil handler means the system should become quiescent, a nil
+// handler means the system should become non-quiescent.  A call
+// to Wait while the system is quiescent will be rebuffed by
+// returning `quiescent=true`.  If all the queues have no requests
+// waiting nor executing while the system is quiescent then the
+// handler will eventually be called with no locks held (even if
+// the system becomes non-quiescent between the triggering state
+// and the required call).
+//
+// The filter uses this for a priority level that has become
+// undesired, setting a handler that will cause the priority level
+// to eventually be removed from the filter if the filter still
+// wants that.  If the filter later changes its mind and wants to
+// preserve the priority level then the filter can use this to
+// cancel the handler registration.
+func (qs *QueueSetImpl) Quiesce(eh EmptyHandler) {
+	qs.lock.Lock()
+	defer qs.lock.Unlock()
+	if eh == nil {
+		qs.quiescent = false
+		return
+	}
+	// Here we check whether there are any requests queued or executing and
+	// if not then fork an invocation of the EmptyHandler.
+	if qs.numPacketsEnqueued == 0 && qs.GetRequestsExecuting() == 0 {
+		// fork an invocation of the EmptyHandler.
+		go func() {
+			eh.HandleEmpty()
+		}()
+	}
+	qs.quiescent = true
```

Comment:
Also need to save the EmptyHandler in `qs`.

Reply:
this was added